Deeply rework WeakIdKeyDictionary

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90825
* #90628

In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Deeply rework WeakIdKeyDictionary

In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Deeply rework WeakIdKeyDictionary"


In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

albanD commented on test/test_weak.py:
```diff
@@ -0,0 +1,570 @@
+        dict2 = WeakIdKeyDictionary(dict)
+        self.assertEqual(dict[o], 364)
+
+    def make_weak_keyed_dict(self):
```
This is not used?

albanD commented on test/test_weak.py:
```diff
@@ -0,0 +1,570 @@
+
+# These tests are ported from cpython/Lib/test/test_weakref.py,
+# but adapted to use tensor rather than object
+class WeakTest(TestCase):
```
There is no test in there that ensures that we only hold a weak reference and not a strong one?
Also any test to ensure we don't introduce ref cycles (which would delay deletion in practice)?

ezyang replied:
Weak reference is indirectly tested, as we do test that entries are promptly removed from dictionary when they die (this is how I found _fix_weakref leak).

There were some ref cycle tests in cpython that seemed complicated, but I could try to intro them

ezyang replied:
leaving this as follow up, I think prompt deletion also covers the ref cycle case

albanD reviewed (approved):
This solves https://github.com/pytorch/pytorch/issues/81597 ?

Sounds good!

eellison commented on torch/utils/weak.py:
```diff
@@ -1,79 +1,263 @@
+        del self.data[WeakIdRef(key)]  # CHANGED
+
+    def __getitem__(self, key):
+        return self.data[WeakIdRef(key)]  # CHANGED
```
we need to do `_fix_weakref` here. You're calling it elsewhere but if we're going to expose this as a utility we shouldn't require users to call it

ezyang replied:
I disagree. I never take out a strong reference to the key here, so I don't need to fix weakref

eellison replied:
You're right. Was thinking that data was a WeakValueDictionary

eellison commented on torch/utils/weak.py:
```diff
@@ -1,79 +1,263 @@
+
+    def pop(self, key, *args):
+        self._dirty_len = True
+        return self.data.pop(WeakIdRef(key), *args)  # CHANGED
```
+ 1, _fix_weakref, right ?

eellison reviewed (commented):
this is not the same as the prior implementation, because the prior implementation was weak on both key and value

ezyang commented:
I don't think so, pretty sure old impl was strong on value

ezyang committed:
Update on "Deeply rework WeakIdKeyDictionary"


In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Deeply rework WeakIdKeyDictionary"


In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "known master failure only"

ezyang committed:
Update on "Deeply rework WeakIdKeyDictionary"


In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang committed:
Update on "Deeply rework WeakIdKeyDictionary"


In the prior patch, I just YOLOed a mutable mapping implementation.
Many edge cases were not handled correctly.  In this PR, I just
copy paste the WeakKeyDictionary from CPython and the hacked it up
to use WeakIdRef instead of weakref.ref.  You can see each line
I changed with the comment CHANGED; there aren't many.

Being exactly API compatible with WeakKeyDictionary means I can also
rob all of the tests from CPython, which I also did for
test/test_weak.py

How to review?  You could either try taking the delta from CPython
(recommended), or review everything from scratch (not recommended).
Can post diff representing delta on request.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "idk doesn't look conflicting to me"

danthe3rd commented on torch/utils/weak.py:
```diff
@@ -1,79 +1,263 @@
+    def __call__(self):
+        r = super().__call__()
+        # Special logic for Tensor PyObject resurrection
+        if hasattr(r, '_fix_weakref'):
+            r._fix_weakref()  # type: ignore[union-attr]
```
This class seems to still keep a strong reference to the object. Repro:
```python
import torch
import gc
import weakref
from _torch_weakkeyid import WeakTensorKeyDictionary, WeakIdRef

x = torch.zeros([1])
y = torch.zeros([1])
refx = weakref.ref(x)
refy = weakref.ref(y)
WeakIdRef(x) == WeakIdRef(torch.ones([1]))
del x
del y
gc.collect()
print("refX: ", type(refx())) # refX:  <class 'torch.Tensor'>
print("refY: ", type(refy())) # refY:  <class 'NoneType'>

```

Commenting out "r._fix_weakref()" here fixes it

ezyang replied:
Do you have the first PR in this stack? There was a memory leak that I fixed

ezyang replied:
yeah on master, fixing your example, it works

```
(/Users/ezyang/Dev/pytorch-cpu-env) ezyang-mbp:pytorch-cpu ezyang$ cat foo.py
import torch
import gc
import weakref
from torch.utils.weak import WeakIdRef

x = torch.zeros([1])
y = torch.zeros([1])
refx = weakref.ref(x)
refy = weakref.ref(y)
WeakIdRef(x) == WeakIdRef(torch.ones([1]))
del x
del y
gc.collect()
print("refX: ", type(refx())) # refX:  <class 'torch.Tensor'>
print("refY: ", type(refy())) # refY:  <class 'NoneType'>
(/Users/ezyang/Dev/pytorch-cpu-env) ezyang-mbp:pytorch-cpu ezyang$ python foo.py
refX:  <class 'NoneType'>
refY:  <class 'NoneType'>
```

danthe3rd replied:
Oh right, thanks a lot! I just imported the `weak.py` file (as I wanted it to work with older versions of pytorch as well). I'll look into the other PR. Thanks!

ezyang replied:
For old versions just get rid of the fix weakref call, it's super busted lol. I suppose if you are willing to write a little C extension code, manually decreffing the Tensor after fixing weakref would work too
